service/history/task/cross_cluster_task.go

```diff
@@ -0,0 +1,385 @@
+		maxRetryCount   dynamicconfig.IntPropertyFn
+	}
+)
+
+// NewCrossClusterSignalWorkflowTask initialize cross cluster signal workflow task and task future
+func NewCrossClusterSignalWorkflowTask(
+	shard shard.Context,
+	taskInfo Info,
+	logger log.Logger,
+	timeSource clock.TimeSource,
+	maxRetryCount dynamicconfig.IntPropertyFn,
+) CrossClusterTask {
+	crossClusterTask := newCrossClusterTask(
+		shard,
+		taskInfo,
+		logger,
+		timeSource,
+		maxRetryCount)
+	return &crossClusterSignalWorkflowTask{
+		crossClusterTask,
+	}
+}
+
+// NewCrossClusterCancelWorkflowTask initialize cross cluster cancel workflow task and task future
+func NewCrossClusterCancelWorkflowTask(
+	shard shard.Context,
+	taskInfo Info,
+	logger log.Logger,
+	timeSource clock.TimeSource,
+	maxRetryCount dynamicconfig.IntPropertyFn,
+) CrossClusterTask {
+	crossClusterTask := newCrossClusterTask(
+		shard,
+		taskInfo,
+		logger,
+		timeSource,
+		maxRetryCount)
+	return &crossClusterCancelWorkflowTask{
+		crossClusterTask,
+	}
+}
+
+// NewCrossClusterStartChildWorkflowTask initialize cross cluster start child workflow task and task future
+func NewCrossClusterStartChildWorkflowTask(
+	shard shard.Context,
+	taskInfo Info,
+	logger log.Logger,
+	timeSource clock.TimeSource,
+	maxRetryCount dynamicconfig.IntPropertyFn,
+) CrossClusterTask {
+	crossClusterTask := newCrossClusterTask(
+		shard,
+		taskInfo,
+		logger,
+		timeSource,
+		maxRetryCount)
+	return &crossClusterStartChildWorkflowTask{
+		crossClusterTask,
+	}
+}
+
+func newCrossClusterTask(
+	shard shard.Context,
+	taskInfo Info,
+	logger log.Logger,
+	timeSource clock.TimeSource,
+	maxRetryCount dynamicconfig.IntPropertyFn,
+) *crossClusterTaskBase {
+	var eventLogger eventLogger
+	if shard.GetConfig().EnableDebugMode &&
+		shard.GetConfig().EnableTaskInfoLogByDomainID(taskInfo.GetDomainID()) {
+		eventLogger = newEventLogger(logger, timeSource, defaultTaskEventLoggerSize)
+		eventLogger.AddEvent("Created task")
+	}
+
+	return &crossClusterTaskBase{
+		Info:          taskInfo,
+		shard:         shard,
+		priority:      ctask.NoPriority,
+		attempt:       0,
+		timeSource:    timeSource,
+		submitTime:    timeSource.Now(),
+		logger:        logger,
+		eventLogger:   eventLogger,
+		scopeIdx:      GetCrossClusterTaskMetricsScope(taskInfo.GetTaskType()),
+		scope:         nil,
+		maxRetryCount: maxRetryCount,
+	}
+}
+
+// Cross cluster signal workflow task
+
+func (c *crossClusterSignalWorkflowTask) Execute() error {
+	panic("Not implement")
+}
+
+func (c *crossClusterSignalWorkflowTask) Ack() {
+	panic("Not implement")
+
+}
+
+func (c *crossClusterSignalWorkflowTask) Nack() {
+	panic("Not implement")
+
+}
+
+func (c *crossClusterSignalWorkflowTask) HandleErr(
+	err error,
+) (retErr error) {
+	panic("Not implement")
+}
+
+func (c *crossClusterSignalWorkflowTask) RetryErr(
+	err error,
+) bool {
+	panic("Not implement")
+}
+
+func (c *crossClusterSignalWorkflowTask) State() ctask.State {
```

Comment:
These common methods (State, Priority, SetPriority, GetShard, GetAttempt, GetQueueType) can be defined only once for `crossClusterTaskBase` I think?

Reply:
Yes. changed to crossClusterTaskBase